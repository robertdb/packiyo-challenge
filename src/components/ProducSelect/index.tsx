export * from './ProducSelect';
export * from './CandidateProductTable';
