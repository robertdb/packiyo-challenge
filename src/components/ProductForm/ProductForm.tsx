'use client';

import { addProduct } from '@/actions/product';
import { Button } from '@/components/ui/button';
import { Input } from '@/components/ui/input';
import { useToast } from '@/hooks/useToast';
import { ReloadIcon } from '@radix-ui/react-icons';
import { Label } from '@radix-ui/react-label';
import { useMemo } from 'react';
import { useFormState, useFormStatus } from 'react-dom';

interface ProductFormProps {
  onCloseDialog: () => void;
}

interface SubmitButtonProps extends ProductFormProps {
  close: boolean;
}

const SubmitButton = ({ close, onCloseDialog }: SubmitButtonProps) => {
  const { pending } = useFormStatus();

  useMemo(() => {
    if (close && !pending) {
      onCloseDialog();
    }
  }, [close, pending]);

  return (
    <Button type="submit" disabled={pending}>
      Save {pending && <ReloadIcon className="ml-2 animate-spin" />}
    </Button>
  );
};

export const ProductForm = ({ onCloseDialog }: ProductFormProps) => {
  const [data, action] = useFormState(addProduct, {});

  useToast(data.error?.server, data?.status === 200 && !data.error ? 'Product Created' : undefined);

  return (
    <form action={action} className="space-y-8">
      <div className="space-y-2">
        <Label htmlFor="name">SKU</Label>
        <Input type="text" id="sku" name="sku" />
        {data?.error?.sku && <div className="text-destructive">{data?.error?.sku}</div>}
      </div>
      <div className="space-y-2">
        <Label htmlFor="name">Name</Label>
        <Input type="text" id="name" name="name" />
        {data?.error?.name && <div className="text-destructive">{data?.error?.name}</div>}
      </div>
      <SubmitButton close={data?.status === 200} onCloseDialog={onCloseDialog} />
    </form>
  );
};
