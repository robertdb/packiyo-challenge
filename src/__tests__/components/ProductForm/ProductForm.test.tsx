import { ProductForm } from '@/components/ProductForm';
import { render } from '@testing-library/react';

jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  useFormState: () => [() => {}, null],
  useFormStatus: () => [() => {}, null],
}));

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn(),
    }));

describe('ProductForm', () => {
  it('matches snapshot', () => {
    const mockOnCloseDialog = jest.fn();
    const { asFragment } = render(<ProductForm onCloseDialog={mockOnCloseDialog} />);
    expect(asFragment()).toMatchSnapshot();
  });
});