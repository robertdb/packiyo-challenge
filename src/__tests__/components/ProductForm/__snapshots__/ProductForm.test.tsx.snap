// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="space-y-8"
  >
    <div
      class="space-y-2"
    >
      <label
        for="name"
      >
        SKU
      </label>
      <input
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id="sku"
        name="sku"
        type="text"
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        for="name"
      >
        Name
      </label>
      <input
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id="name"
        name="name"
        type="text"
      />
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      type="submit"
    >
      Save 
    </button>
  </form>
</DocumentFragment>
`;
